crates/re_types_builder/src/codegen/cpp.rs

```diff
@@ -0,0 +1,175 @@
+                write_file(&filepath, string);
+                let inserted = filepaths.insert(filepath);
+                assert!(
+                    inserted,
+                    "Multiple objects with the same name: {:?}",
+                    obj.name
+                );
+            }
+        }
+
+        {
+            // Generate module file that includes all the headers:
+            let hash = quote! { # };
+            let pragma_once = pragma_once();
+            let header_file_names = ordered_objects
+                .iter()
+                .map(|obj| format!("{folder_name}/{}.hpp", obj.snake_case_name()));
+            let tokens = quote! {
+                #pragma_once
+                #(#hash include #header_file_names "RE_TOKEN_NEWLINE")*
+            };
+            let filepath = folder_path
+                .parent()
+                .unwrap()
+                .join(format!("{folder_name}.hpp"));
+            let string = string_from_token_stream(&tokens, None);
+            write_file(&filepath, string);
+            filepaths.insert(filepath);
+        }
+
+        // Clean up old files:
+        for entry in std::fs::read_dir(folder_path).unwrap().flatten() {
+            let filepath = Utf8PathBuf::try_from(entry.path()).unwrap();
+            if !filepaths.contains(&filepath) {
+                std::fs::remove_file(filepath).ok();
+            }
+        }
+
+        filepaths
+    }
+}
+
+impl crate::CodeGenerator for CppCodeGenerator {
+    fn generate(
+        &mut self,
+        objects: &Objects,
+        arrow_registry: &ArrowRegistry,
+    ) -> BTreeSet<Utf8PathBuf> {
+        ObjectKind::ALL
+            .par_iter()
+            .map(|object_kind| {
+                let folder_name = object_kind.plural_snake_case();
+                self.generate_folder(objects, arrow_registry, *object_kind, folder_name)
+            })
+            .flatten()
+            .collect()
+    }
+}
+
+fn string_from_token_stream(token_stream: &TokenStream, source_path: Option<&Utf8Path>) -> String {
+    let mut code = String::new();
+    code.push_str(&format!("// {AUTOGEN_WARNING}\n"));
+    if let Some(source_path) = source_path {
+        code.push_str(&format!("// Based on {source_path:?}"));
+    }
+
+    code.push('\n');
+    code.push_str(
+        &token_stream
+            .to_string()
+            .replace(&format!("{NEWLINE_TOKEN:?}"), "\n"),
+    );
+    code.push('\n');
+
+    // clang_format has a bit of an ugly API: https://github.com/KDAB/clang-format-rs/issues/3
+    clang_format::CLANG_FORMAT_STYLE
+        .set(clang_format::ClangFormatStyle::File)
+        .ok();
+    code = clang_format::clang_format(&code).expect("Failed to run clang-format");
+
+    code
+}
+
+fn write_file(filepath: &Utf8PathBuf, code: String) {
+    if let Ok(existing) = std::fs::read_to_string(filepath) {
+        if existing == code {
+            // Don't touch the timestamp unnecessarily
+            return;
```

Comment:
any reason we do this here and not in the other generators? another candidate for becoming a general utility

Reply:
the reason is I didn't write the other generators :P